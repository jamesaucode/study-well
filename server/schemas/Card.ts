import { Document, Schema, Model, model } from 'mongoose';

type UserId = string;

export interface ICard {
    question: string,
    answer: string,
    creator: UserId 
}
export interface ICardModel extends Document, ICard {
    findByUserId(id : UserId): boolean;
}

export const CardSchema : Schema<ICardModel> = new Schema({
    question: String,
    answer: String,
    creator: Schema.Types.ObjectId 
})

CardSchema.statics.findByUserId = function(id : UserId) : boolean {
    return this.find({ creator: id });
}

export const CardsModel : Model<ICardModel> = model<ICardModel>('Cards', CardSchema);