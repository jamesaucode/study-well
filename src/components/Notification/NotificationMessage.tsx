import React from "react";
import styled from "styled-components";
import { Message } from "../../styles/shared";
import { NextFC } from "next";
import { IMessage } from "../../../interfaces/message";
import SVG from "react-inlinesvg";

interface WrapperProps {
  mounted?: boolean;
}
const Wrapper = styled.div<WrapperProps>`
  display: flex;
  justify-content: center;
  border-radius: 5px;
  width: 100%;
  padding: 0.75rem;
  box-sizing: border-box;
  transition: 0.5s ease-in-out opacity;
`;
const Important = styled.span`
  font-weight: 700;
`;
const SVGWrapper = styled.div`
  padding: 0 0.5rem;
`;
const renderImportant = (text: string): React.ReactElement => (
  <Important>{text}</Important>
);
const NotificationMessage: NextFC<IMessage> = ({
  success,
  message,
  id,
  // delay = 3000,
  removeNotification,
}): JSX.Element => {
  return (
    <Wrapper key={id}>
      <Message success={success}>
        <div>
          {success ? renderImportant("Success! ") : renderImportant("Oops! ")}
          {message}
        </div>
        <SVGWrapper onClick={(): void => removeNotification(id)}>
          <SVG className="small-icon" src="/static/images/close.svg" />
        </SVGWrapper>
      </Message>
    </Wrapper>
  );
};

NotificationMessage.defaultProps = {
  success: false,
  message: "No message",
};

export default React.memo(NotificationMessage);
