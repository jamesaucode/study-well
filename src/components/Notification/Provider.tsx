import React, { useRef, useState, useEffect } from 'react';
import { Context } from './Context';

const Provider = () => {
    const root  = useRef<Node | null>(null);
    const id = useRef([]);
    const [notifications, setNotifications] = useState([]);

    useEffect(() => {
        root.current = document.createElement('div');
        document.body.appendChild(root.current);
        return () => {
           id.current.forEach(clearTimeout);
           if (root.current) {
            document.body.removeChild(root.current);
           }
        };
    }, [])

    // const remove = (notification) => {

    // }
}
