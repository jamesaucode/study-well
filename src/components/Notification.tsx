import React, { useState, useEffect, useRef } from "react";
import NotificationMessage from "./NotificationMessage";
import { MessageType } from "../../typings/message";

const Notification = (props: any) => {
    const [notifications, setNotifications] = useState();
    const popNotification = () => {
        setNotifications(notifications.slice(0, -1));
    }
    const pushNotification = (message  = "lmaooooo", status: boolean) => {
        setNotifications(notifications.concat({
            message,
            success: status
        }))
    }

  const children = React.Children.map(props.children, child => {
      return React.cloneElement(child, {
          popNotification,
          pushNotification 
      })
  })
  if (notifications) {
    return (
      <React.Fragment>
        {notifications.map((notification: MessageType) => {
          return (
            <NotificationMessage
              message={notification.message}
              success={notification.success}
            />
          );
        })}
        {children}
      </React.Fragment>
    );
  } else {
    return <button onClick={popNotification}>Test Button</button>;
  }
};

export default React.memo(Notification);
