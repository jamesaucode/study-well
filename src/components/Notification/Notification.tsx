import React, { useState, useEffect, useRef, useMemo } from "react";
import NotificationMessage from "./NotificationMessage";
import { MessageType } from "../../../typings/message";
import styled from "styled-components";

const uuid = require("uuid");

const AllNotificationsWrapper = styled.div`
  height: 0; // Hacky ?
  top: 20px;
  position: sticky;
  z-index:10;
`;
const Notification = (props: any) => {
  const [notifications, setNotifications] = useState<MessageType[]>([]);
  const delay = 5000;
  const id = useRef<number[]>([]);
  useEffect(() => {
    return () => {
      id.current.forEach(clearTimeout);
    };
  }, []);
  const removeNotification = (notificationId: string) => {
    setNotifications(prevState => {
      console.log("Removing Notification");
      return prevState.filter(n => n.id !== notificationId);
    });
  };
  const pushNotification = (message = "Lol", status: boolean) => {
    const notification = {
      message,
      success: status,
      id: uuid(),
      delay,
    };
    console.log("Setting up the timer");
    const timerId = setTimeout(() => {
      console.log("Now removing the notification");
      removeNotification(notification.id);
      id.current.splice(id.current.indexOf(timerId), 1);
    }, delay);
    id.current.push(timerId);
    setNotifications(allNotification => allNotification.concat(notification));
    return notification;
  };

  const children = React.Children.map(props.children, child => {
    return React.cloneElement(child, {
      pushNotification,
      removeNotification,
    });
  });
  if (notifications) {
    return (
      <React.Fragment>
        <AllNotificationsWrapper>
          {notifications.map((notification: MessageType) => {
            return (
              <NotificationMessage
                message={notification.message}
                success={notification.success}
                id={notification.id}
                delay={notification.delay}
                removeNotification={removeNotification}
              />
            );
          })}
        </AllNotificationsWrapper>
        {children}
      </React.Fragment>
    );
  } else {
    return <React.Fragment>{children}</React.Fragment>;
  }
};

export default React.memo(Notification);
