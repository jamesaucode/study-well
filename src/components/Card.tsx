import React, { useState } from "react";
import styled, { keyframes } from "styled-components";
import { font, colors } from "../../utils/style";
import { useNotification } from "./Notification/Notification";

interface PropTypes {
  _id: string;
  question: string;
  answer: string;
  creator: string;
  controls: {
    nextCard: () => void;
    previousCard: () => void;
  };
}
const slideIn = keyframes`
  from {
    transform: translateX(-10%);
    opacity: 0;
  }
  to {
    transform: translateX(0);
    opacity: 1;
  }
`;
const CardBody = styled.div`
  display: flex;
  flex: 1;
  flex-direction: column;
  justify-content: space-between;
`;
const CardText = styled.h2<{ shouldSlideIn?: boolean }>`
  color: ${colors.black};
  font-size: ${font.fontSize.lg};
  animation: ${slideIn} 400ms ease 1 forwards;
`;
const AnswerInput = styled.input`
  padding: 0.5em;
  background-color: transparent;
  border: none;
  border-bottom: 1px solid gray;
`;

const Card: React.FunctionComponent<PropTypes> = ({
  _id,
  question,
  answer,
  controls,
}): JSX.Element => {
  const [currentAnswer, setCurrentAnswer] = useState("");
  const [showAnswer, setShowAnswer] = useState(false);
  const { pushNotification } = useNotification();
  const handleKeyDown = (
    event: React.KeyboardEvent<HTMLInputElement>,
  ): void => {
    if (event.keyCode === 13) {
      if (currentAnswer === answer) {
        pushNotification("Correct!", true);
        controls.nextCard();
        setCurrentAnswer("");
      } else {
        setShowAnswer(true);
      }
    }
  };
  const handleChange = (event: React.ChangeEvent<HTMLInputElement>): void => {
    setCurrentAnswer(event.target.value);
  };
  return (
    <CardBody>
      <CardText>{question}</CardText>
      {showAnswer && <CardText>{answer}</CardText>}
      <AnswerInput
        onChange={handleChange}
        placeholder="Answer"
        name="answer"
        value={currentAnswer}
        onKeyDown={handleKeyDown}
      />
    </CardBody>
  );
};

export default Card;
