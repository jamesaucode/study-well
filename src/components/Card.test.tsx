/* eslint-disable */
import React from "react";
import { mount } from "enzyme";
import Card from "./Card";
import { NotificationProvider } from "./Notification/Notification";

const defaultProps = {
  _id: "",
  question: "",
  answer: "",
  creator: "",
  controls: {
    nextCard: (): void => {},
    previousCard: (): void => {},
  },
};
describe("Card Component", () => {
  it("Render Component", () => {
    const wrapper = mount(
      <NotificationProvider>
        <Card {...defaultProps} />
      </NotificationProvider>,
    );
    expect(wrapper).toMatchSnapshot();
  });
});
