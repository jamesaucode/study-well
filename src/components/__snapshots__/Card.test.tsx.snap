// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card Component Render Component 1`] = `
<Card
  _id=""
  answer=""
  controls={
    Object {
      "nextCard": [Function],
      "previousCard": [Function],
    }
  }
  creator=""
  pushNotification={[Function]}
  question=""
>
  <Card__CardBody>
    <StyledComponent
      forwardedComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "Card__CardBody-d2183a-0",
            "isStatic": true,
            "lastClassName": "bnnVid",
            "rules": Array [
              "display:flex;flex:1;flex-direction:column;justify-content:space-between;",
            ],
          },
          "displayName": "Card__CardBody",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "Card__CardBody-d2183a-0",
          "target": "div",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={null}
    >
      <div
        className="Card__CardBody-d2183a-0 bnnVid"
      >
        <Card__CardText
          shouldSlideIn={true}
        >
          <StyledComponent
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "Card__CardText-d2183a-1",
                  "isStatic": true,
                  "lastClassName": "gDeGUy",
                  "rules": Array [
                    "color:",
                    "#222222",
                    ";font-size:",
                    "1.5em",
                    ";transition:transform 400ms ease;",
                  ],
                },
                "displayName": "Card__CardText",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "Card__CardText-d2183a-1",
                "target": "h2",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
            shouldSlideIn={true}
          >
            <h2
              className="Card__CardText-d2183a-1 gDeGUy"
            />
          </StyledComponent>
        </Card__CardText>
        <Card__AnswerInput
          name="answer"
          onChange={[Function]}
          onKeyDown={[Function]}
          placeholder="Answer"
          value=""
        >
          <StyledComponent
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "Card__AnswerInput-d2183a-2",
                  "isStatic": true,
                  "lastClassName": "jwEAoi",
                  "rules": Array [
                    "padding:0.5em;background-color:transparent;border:none;border-bottom:1px solid gray;",
                  ],
                },
                "displayName": "Card__AnswerInput",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "Card__AnswerInput-d2183a-2",
                "target": "input",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
            name="answer"
            onChange={[Function]}
            onKeyDown={[Function]}
            placeholder="Answer"
            value=""
          >
            <input
              className="Card__AnswerInput-d2183a-2 jwEAoi"
              name="answer"
              onChange={[Function]}
              onKeyDown={[Function]}
              placeholder="Answer"
              value=""
            />
          </StyledComponent>
        </Card__AnswerInput>
      </div>
    </StyledComponent>
  </Card__CardBody>
</Card>
`;
